### ISSUE TITLE
As a software developer of scientific software, I need a native Haskell implementation of the daxpy function so that I can linearly combine two vectors in a type-safe, thread safe manner.

### ISSUE THREAD

**Issue by dlewissandy** [dlewissandy/lambda-blas#72]
Documentation for the daxpy function can be found at BLAS.

- [x] create native implementation of function
- [x] create haddock documentation of function
- [x] create unit test to demonstrate byte equivalence to FORTRAN BLAS library function
- [x] create benchmark(s) to quantify time performance relative to FORTRAN BLAS library function
- [x] demonstrate that native has a relative time performance of no more than 100x the FORTRAN blas function.

**Comment by dlewissandy** [comment]
Implementation Strategy:
* This function can be implemented easily by polymorphising the single precision version.

**Comment by dlewissandy** [comment]
Attached is the raw benchmark data:
[daxpy.zip](https://github.com/dlewissandy/lambda-blas/files/868045/daxpy.zip)

**Comment by dlewissandy** [comment]
# DAXPY BENCHMARK ANALYSIS #

## Special Case inc=1 ##
When the increment of traversal is equal to unity, the time performance of both the unsafe and pure implementations are well described by a linear model:

### PURE ###
```
lm(formula = PURE ~ N, data = dat[dat$INC == 1, ])

Residuals:
     Min       1Q   Median       3Q      Max
-0.60543 -0.07155  0.06697  0.07570  0.65397

Coefficients:
             Estimate Std. Error t value Pr(>|t|)
(Intercept) 2.421e-02  4.186e-02   0.578    0.566
N           5.804e-03  6.362e-06 912.236   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.2362 on 37 degrees of freedom
Multiple R-squared:      1,	Adjusted R-squared:      1
F-statistic: 8.322e+05 on 1 and 37 DF,  p-value: < 2.2e-16
```
### UNSAFE ###
```
lm(formula = UNSAFE ~ N, data = dat[dat$INC == 1, ])

Residuals:
    Min      1Q  Median      3Q     Max
-1.1030 -0.1527  0.2170  0.2522  0.9794

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -1.839e-01  8.380e-02  -2.194   0.0346 *
N            1.101e-03  1.274e-05  86.476   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.4728 on 37 degrees of freedom
Multiple R-squared:  0.9951,	Adjusted R-squared:  0.9949
F-statistic:  7478 on 1 and 37 DF,  p-value: < 2.2e-16
```
### ASYMPTOTIC PERFORMANCE ###
In the limit as n becomes very large, the relative time performance should approach:
```
5.804e-03
--------- = 5.2x
1.101e-03
```
which is within the 100x performance threshold for this project milestone

## Special Case inc/=1 ##
When the increment of traversal is not equal to unity, the time performance of both the unsafe and pure implementations are well described by a multi-linear model:

### PURE ###
```
lm(formula = PURE ~ N + Z, data = dat[dat$INC != 1, ])

Residuals:
    Min      1Q  Median      3Q     Max
-0.8875 -0.2509  0.1208  0.2933  0.8693

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -1.690e-01  6.475e-02   -2.61    0.012 *
N            1.513e-02  1.511e-04  100.15   <2e-16 ***
Z            7.696e-04  1.155e-05   66.66   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3936 on 48 degrees of freedom
Multiple R-squared:  0.999,	Adjusted R-squared:  0.999
F-statistic: 2.519e+04 on 2 and 48 DF,  p-value: < 2.2e-16
```
### UNSAFE ###
```
lm(formula = UNSAFE ~ N + Z, data = dat[dat$INC != 1, ])

Residuals:
    Min      1Q  Median      3Q     Max
-1.3590 -0.1767  0.1472  0.2908  1.1798

Coefficients:
              Estimate Std. Error t value Pr(>|t|)
(Intercept) -2.535e-01  8.256e-02  -3.071  0.00351 **
N            5.493e-03  1.926e-04  28.518  < 2e-16 ***
Z            4.533e-04  1.472e-05  30.792  < 2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.5018 on 48 degrees of freedom
Multiple R-squared:  0.9925,	Adjusted R-squared:  0.9922
F-statistic:  3165 on 2 and 48 DF,  p-value: < 2.2e-16
```
### ASYMPTOTIC PERFORMANCE ###
In the limit as both n and inc become very large, the relative time performance should approach:
```
7.696e-04
--------- = 1.7x
4.533e-04
```
which is within the 100x performance threshold for this project milestone.

**Comment by dlewissandy** [comment]
Closed as DONE after approval of pull request #80.
